### PR TITLE
Fix #599 by ignoring merges of config object and primitive types

### LINKED ISSUE #599: Some substitutions are resolved too early
I have created a minimal Scala 2.12 project with no dependencies and one test case that simply resolves the config files and prints the result:
```scala
class ConfigSpec extends FlatSpec with Matchers {
  "The config" should "resolve" in {
    val conf: Config = ConfigFactory.load()
    println(conf.root().render(
      ConfigRenderOptions.concise().setFormatted(true)
    ))
  }
}
```

I found two (probably related) errors in which a substitution gets resolved too early. In both cases, I have an application.conf, but no reference.conf.

1. With the following application.conf:
```
foo = ""
foo {
  bar = ${override-me}
}
foo {
  bar = "bar"
}
```
I get the following exception:
```
com.typesafe.config.ConfigException$UnresolvedSubstitution: application.conf @ file:<redacted>/target/scala-2.12/classes/application.conf: 3: Could not resolve substitution to a value: ${override-me}
```
I get the same exception if I change the first line to `foo = null`, `foo = 0`, etc. Notably, `foo = {}` works just fine. I thought of a "fix": if I add the following object after the first line, then the config resolves correctly:
```
foo {
  bar = ""
}
```

2. With the following application.conf, I get the same exception as above (with a different line number, of course):
```
empty = {}
foo = ${empty}
foo {
  bar = ${override-me}
}
foo {
  bar = "bar"
}
```
The "fix" from the first case doesn't seem to work here.

### PR DESCRIPTION
**Pull request by FingolfinTEK**
Left comments regarding the solution in the issue itself.

### CODE REVIEW

File: config/src/test/scala/com/typesafe/config/impl/ConfigTest.scala [hunk 1]
```diff
@@ -252,20 +252,16 @@ class ConfigTest extends TestUtils {
...
-        }
-
-        associativeMerge(Seq(obj3, obj2, obj1)) { unresolved =>
-            val merged2 = resolveNoSystem(unresolved, unresolved)
-            assertEquals(43, merged2.getInt("a.b"))
-            assertEquals(44, merged2.getInt("a.c"))
-            assertEquals(4, merged2.root.size)
-            assertEquals(2, merged2.getObject("a").size())
-        }
+        val merged = obj1.withFallback(obj2.withFallback(obj3)).toConfig.resolve()
```
[the remark this answers is not in the document]
Comment (havocp) [reply config/src/test/scala/com/typesafe/config/impl/ConfigTest.scala:255]: I think the fix might need to be at the resolve() level?

[the remark this answers is not in the document]
Comment (FingolfinTEK) [reply config/src/test/scala/com/typesafe/config/impl/ConfigTest.scala:255]: Hi @havocp, thanks for your feedback. The thing is - the same mechanism that's allowing for associativity is what's causing the bug i.e. the delayed merge resolution, A non-object merged with an object produces a delayed merge and the delayed merge tries to resolve all elements in the stack, even when it shouldn't, like in the case of `com.typesafe.config.impl.ConfigSubstitutionTest#ignoreHiddenUndefinedSubstOfDifferentType`. 

The main reason I created the pull request is to get feedback and suggestions - the approach you suggest had me changing the `ConfigDelayedMerge` but what I've found is that I couldn't change the `ConfigDelayedMerge` in a way that fixed the reported issue and also didn't break a fair deal of other tests, so I went with a solution that breaks the associativity in that one particular case.

The only other idea I had is to just swallow `UnresolvedSubstitution` errors in delayed merges if the merged object already contained all the keys from the unresolved object but I was unsure if that could create a new set of issues down the line (all the tests pass with that change).

[the remark this answers is not in the document]
Comment (FingolfinTEK) [reply config/src/test/scala/com/typesafe/config/impl/ConfigTest.scala:255]: Actually, let me push that last one so you can get a better idea of what I'm talking about.

File: config/src/main/java/com/typesafe/config/impl/ConfigDelayedMerge.java [hunk 2]
```diff
@@ -129,9 +129,18 @@ else if (end instanceof Unmergeable) {
...
                 ConfigImpl.trace(newContext.depth(), "Resolving highest-priority item in delayed merge " + end
                         + " against " + sourceForEnd + " endWasRemoved=" + (source != sourceForEnd));
-            ResolveResult<? extends AbstractConfigValue> result = newContext.resolve(end, sourceForEnd);
-            AbstractConfigValue resolvedEnd = result.value;
-            newContext = result.context;
+
+
+            AbstractConfigValue resolvedEnd = null;
+            try {
+                ResolveResult<? extends AbstractConfigValue> result = newContext.resolve(end, sourceForEnd);
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigDelayedMerge.java:136]: The bug as I understand it is that here we're resolving more than we need to in 
order to eliminate the delayed merge. We really only need to resolve one "level" and then merge the values together.
We don't need to resolve recursively.

I came up with another fix in https://github.com/lightbend/config/pull/697 that fixes the immediate bug by avoiding the situation where we have two objects here that can be merged without having to resolve them, but my fix may not handle the case where two substitutions in the delayed merge resolve to objects.

Comment (FingolfinTEK) [reply config/src/main/java/com/typesafe/config/impl/ConfigDelayedMerge.java:136]: That definitely looks like a better solution to me, closing the PR :)

**Pull request closed without merging**